node/router/src/heartbeat.rs

```diff
@@ -182,50 +203,36 @@ pub trait Heartbeat<N: Network>: Outbound<N> {
                 .into_iter()
-                .filter(|peer_ip| !trusted.contains(peer_ip) && !bootstrap.contains(peer_ip))
+                .filter(|peer_addr| !trusted.contains(peer_addr) && !bootstrap.contains(peer_addr))
                 .choose_multiple(rng, num_surplus_provers);
 
-            // TODO (howardwu): As a validator, prioritize disconnecting from clients.
-            //  Remove RNG, pick the `n` oldest nodes.
             // Determine the clients and validators to disconnect from.
-            let peer_ips_to_disconnect = self
-                .router()
-                .get_connected_peers()
+            let peers_to_disconnect = self
+                .get_removable_peers()
                 .into_iter()
-                .filter_map(|peer| {
-                    let peer_ip = peer.ip();
-                    if !peer.is_prover() && // Skip if the peer is a prover.
-                       !trusted.contains(&peer_ip) && // Skip if the peer is trusted.
-                       !bootstrap.contains(&peer_ip) && // Skip if the peer is a bootstrap peer.
-                       // Skip if you are syncing from this peer.
-                       (self.is_block_synced() || (!self.is_block_synced() && self.router().cache.num_outbound_block_requests(&peer.ip()) == 0))
-                    {
-                        Some(peer_ip)
-                    } else {
-                        None
-                    }
-                })
-                .choose_multiple(rng, num_surplus_clients_validators);
+                .filter(|peer| !peer.is_prover()) // remove provers as those are handled seperately
+                .map(|p| p.ip())
+                .take(num_surplus_clients_validators);
 
             // Proceed to send disconnect requests to these peers.
-            for peer_ip in peer_ips_to_disconnect.into_iter().chain(prover_ips_to_disconnect) {
+            for peer_addr in peers_to_disconnect.chain(provers_to_disconnect) {
                 // TODO (howardwu): Remove this after specializing this function.
                 if self.router().node_type().is_prover() {
-                    if let Some(peer) = self.router().get_connected_peer(&peer_ip) {
+                    if let Some(peer) = self.router().get_connected_peer(&peer_addr) {
                         if peer.node_type().is_validator() {
                             continue;
                         }
                     }
                 }
 
-                info!("Disconnecting from '{peer_ip}' (exceeded maximum connections)");
-                self.send(peer_ip, Message::Disconnect(DisconnectReason::TooManyPeers.into()));
+                info!("Disconnecting from '{peer_addr}' (exceeded maximum connections)");
+                self.send(peer_addr, Message::Disconnect(DisconnectReason::TooManyPeers.into()));
                 // Disconnect from this peer.
-                self.router().disconnect(peer_ip);
+                self.router().disconnect(peer_addr);
```

Comment:
I think `peer_ip` here was correct? `disconnect` internally calls `resolve_to_ambiguous` which maps the listener to the ephemeral address (ip + port). Granted `peer_ip` is a misnomer but it represents the peer's listener address (also ip + port).

Reply:
We could add some more documentation to `Resolver`. It should also mention that the peer_id / listen address is used to uniquely identify a peer in some parts of the code.

Reply:
Can be done in a separate PR.